pkg/hive/cell/health.go

```diff
@@ -0,0 +1,249 @@
+// waiting for the queue to be processed.
+const updatesBufferSize = 2048
+
+// Level denotes what kind an update is.
+type Level string
+
+const (
+	StatusUnknown  Level = "Unknown"
+	StatusStopped  Level = "Stopped"
+	StatusDegraded Level = "Degraded"
+	StatusOK       Level = "OK"
+)
+
+// StatusReporter provides a method of declaring a Modules health status.
+type StatusReporter interface {
+	// OK declares that a Module has achieved a desired state and has not entered
+	// any unexpected or incorrect states.
+	// Modules should only declare themselves as 'OK' once they have stabilized,
+	// rather than during their initial state. This should be left to be reported
+	// as the default "unknown" to denote that the module has not reached a "ready"
+	// health state.
+	OK(status string)
+
+	// Stopped reports that a module has completed, and will no longer report any
+	// health status.
+	Stopped(reason string)
+
+	// Degraded declares that a module has entered a degraded state.
+	// This means that it may have failed to provide it's intended services, or
+	// to perform it's desired task.
+	Degraded(reason string, err error)
+}
+
+// Health provides exported functions for accessing health status data.
+// As well, provides unexported functions for use during module apply.
+type Health interface {
+	// All returns a copy of all module statuses.
+	// This includes unknown status for modules that have not reported a status yet.
+	All() []Status
+
+	// Get returns a copy of a modules status, by module ID.
+	// This includes unknown status for modules that have not reported a status yet.
+	Get(string) *Status
+
+	// forModule creates a moduleID scoped reporter handle.
+	forModule(string) StatusReporter
+}
+
+// Update is an event that denotes the change of a modules health state.
+type Update struct {
+	Level
+	ModuleID string
+	Message  string
+	Err      error
+}
+
+// Status is a modules last health state, including the last update.
+type Status struct {
+	Update
+	Stopped     bool
+	LastOK      time.Time
+	LastUpdated time.Time
+}
+
+func (s *Status) String() string {
+	var sinceLast string
+	if s.LastUpdated.IsZero() {
+		sinceLast = "never"
+	} else {
+		sinceLast = time.Since(s.LastUpdated).String() + " ago"
+	}
+	return fmt.Sprintf("Status{ModuleID: %s, Level: %s, Since: %s, Message: %s, Err: %v}",
+		s.ModuleID, s.Level, sinceLast, s.Message, s.Err)
+}
+
+// NewHealthProvider starts and returns a health status which collects emitted
+// health status declarations and updates module health state.
+func NewHealthProvider() *HealthProvider {
+	p := &HealthProvider{
+		updates:        make(chan Update, updatesBufferSize),
+		moduleStatuses: make(map[string]Status),
+	}
+	p.running.Store(true)
+
+	p.obs, p.emit, p.complete = stream.Multicast[Update]()
+
+	// Listen for updates, use buffered channel to avoid blocking.
+	go func() {
+		for u := range p.updates {
+			prev := func() Status {
```

Comment:
Why are declaring a closure and then calling it immediately?

Reply:
Not to mention the potential for gor leak if Stop never hits. Better to fire this off on start with a context to ensure we clean up?

Reply:
I use this to capture the critical section with a `defer mu.Unlock()` (i.e. makes it less likely someone will accidentally introduce some logic that leaves the lock stuck), rather than having an explicit unlock.